Source/MockDefaultValueProvider.cs

```diff
@@ -59,23 +60,40 @@ public MockDefaultValueProvider(Mock owner)
 
 		public override object ProvideDefault(MethodInfo member)
 		{
-			var value = base.ProvideDefault(member);
+			Type taskResultType = null;
 
-			Mock mock = null;
+			var returnType = member.ReturnType;
+			if (returnType.GetTypeInfo().IsGenericType)
+			{
+				var returnTypeGenericDef = returnType.GetGenericTypeDefinition();
+				if (returnTypeGenericDef == typeof(Task<>))
+				{
+					taskResultType = returnType.GetGenericArguments()[0];
+				}
+			}
+
+			object value = null;
+			if (taskResultType == null)
```

Comment:
This is still pretty ugly. What these lines of code basically amount to is, "If I have some special handling capabilities for some special type (`Task<T>`), don't let the fallback value provider generate a value for it; I'll deal with it myself right away." Find a way to express this more clearly in code.